File: backend/app/models/attendance.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import String, DateTime, Date, ForeignKey, Boolean, Float, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class AttendanceRecord(Base):
    __tablename__ = "attendance_records"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    user_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False
    )
    record_date: Mapped[str] = mapped_column(String(10), nullable=False)  # "2026-03-27"
    record_type: Mapped[str] = mapped_column(String(10), nullable=False)  # "checkin" | "checkout"

    # 验证结果
    face_verified: Mapped[bool] = mapped_column(Boolean, default=False)
    face_score: Mapped[float | None] = mapped_column(Float, nullable=True)
    location_verified: Mapped[bool | None] = mapped_column(Boolean, nullable=True)
    latitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    longitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    distance_m: Mapped[float | None] = mapped_column(Float, nullable=True)

    # 时间状态
    is_late: Mapped[bool] = mapped_column(Boolean, default=False)
    is_early_leave: Mapped[bool] = mapped_column(Boolean, default=False)

    # 审计
    device_info: Mapped[str | None] = mapped_column(String(255), nullable=True)
    photo_path: Mapped[str | None] = mapped_column(String(500), nullable=True)

    recorded_at: Mapped[datetime] = mapped_column(
        DateTime, default=lambda: datetime.now(timezone.utc), server_default=func.now()
    )

    user = relationship("User", back_populates="attendance_records")

File: mobile/lib/models/attendance.dart
class AttendanceRecord {
  final String id;
  final String userId;
  final String? userName;
  final String recordDate;
  final String recordType;
  final bool faceVerified;
  final double? faceScore;
  final bool? locationVerified;
  final double? latitude;
  final double? longitude;
  final double? distanceM;
  final bool isLate;
  final bool isEarlyLeave;
  final String? deviceInfo;
  final String recordedAt;

  AttendanceRecord({
    required this.id,
    required this.userId,
    this.userName,
    required this.recordDate,
    required this.recordType,
    required this.faceVerified,
    this.faceScore,
    this.locationVerified,
    this.latitude,
    this.longitude,
    this.distanceM,
    required this.isLate,
    required this.isEarlyLeave,
    this.deviceInfo,
    required this.recordedAt,
  });

  factory AttendanceRecord.fromJson(Map<String, dynamic> json) {
    return AttendanceRecord(
      id: json['id'],
      userId: json['user_id'],
      userName: json['user_name'],
      recordDate: json['record_date'],
      recordType: json['record_type'],
      faceVerified: json['face_verified'] ?? false,
      faceScore: json['face_score']?.toDouble(),
      locationVerified: json['location_verified'],
      latitude: json['latitude']?.toDouble(),
      longitude: json['longitude']?.toDouble(),
      distanceM: json['distance_m']?.toDouble(),
      isLate: json['is_late'] ?? false,
      isEarlyLeave: json['is_early_leave'] ?? false,
      deviceInfo: json['device_info'],
      recordedAt: json['recorded_at'],
    );
  }

  String get typeLabel => recordType == 'checkin' ? '签到' : '签退';

  String get statusLabel {
    if (isLate) return '迟到';
    if (isEarlyLeave) return '早退';
    return '正常';
  }
}
